packages/core/streams/langchain-stream.test.ts

```diff
@@ -10,107 +29,150 @@ describe('LangchainStream', () => {
-        '0:""\n',
-        '0:"Hello"\n',
-        '0:","\n',
-        '0:" world"\n',
-        '0:"."\n',
-        '0:""\n',
+        '',
+        'Hello',
+        ',',
+        ' world',
+        '.',
+        '',
       ]);
     });
 
-    it('should send text and data', async () => {
-      const data = new experimental_StreamData();
+    describe('StreamData prototcol', () => {
+      it('should send text', async () => {
+        const data = new experimental_StreamData();
 
-      data.append({ t1: 'v1' });
+        const model = new ChatOpenAI({
+          streaming: true,
+          openAIApiKey: 'fake',
+          configuration: {
+            baseURL: server.api,
+            defaultHeaders: {
+              'x-mock-service': 'openai',
+              'x-mock-type': 'chat',
+            },
+          },
+        });
 
-      const { stream, handlers } = LangChainStream({
-        onFinal() {
-          data.close();
-        },
-        experimental_streamData: true,
+        const stream = await PromptTemplate.fromTemplate('{input}')
+          .pipe(model)
+          .pipe(new BytesOutputParser())
+          .stream(
+            { input: 'Hello' },
+            {
+              callbacks: [
+                {
+                  handleChainEnd(outputs, runId, parentRunId) {
+                    // check that main chain (without parent) is finished:
+                    if (parentRunId == null) {
+                      data.close();
+                    }
+                  },
+                },
+              ],
+            },
+          );
+
+        const response = new StreamingTextResponse(
+          stream.pipeThrough(createStreamDataTransformer(true)),
+          {},
+          data,
+        );
+
+        expect(await readAllChunks(response)).toEqual([
+          '0:""\n',
+          '0:"Hello"\n',
+          '0:","\n',
+          '0:" world"\n',
+          '0:"."\n',
+          '0:""\n',
+        ]);
+      });
+
+      it('should send text and data', async () => {
+        const data = new experimental_StreamData();
+
+        data.append({ t1: 'v1' });
+
+        const model = new ChatOpenAI({
+          streaming: true,
+          openAIApiKey: 'fake',
+          configuration: {
+            baseURL: server.api,
+            defaultHeaders: {
+              'x-mock-service': 'openai',
+              'x-mock-type': 'chat',
+            },
+          },
+        });
+
+        const stream = await PromptTemplate.fromTemplate('{input}')
+          .pipe(model)
+          .pipe(new BytesOutputParser())
+          .stream(
+            { input: 'Hello' },
+            {
+              callbacks: [
+                {
+                  handleChainEnd(outputs, runId, parentRunId) {
+                    // check that main chain (without parent) is finished:
+                    if (parentRunId == null) {
+                      data.close();
+                    }
+                  },
+                },
+              ],
```

Comment:
This is the main trick that's required for closing `data` correctly.

Reply:
hm, not ideal but we can document this clearly. maybe we can handle it for the user post-experimental.